hal/sparcv8leon3/gaisler/timer.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Timer controller
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include "hal/timer.h"
+#include "hal/spinlock.h"
+
+#ifdef NOMMU
+#define VADDR_GPTIMER0 GPTIMER0_BASE
+#else
+#define VADDR_GPTIMER0 (void *)((u32)VADDR_PERIPH_BASE + PAGE_OFFS_GPTIMER0)
+#endif
+
+/* Timer control bitfields */
+
+#define TIMER_ENABLE      (1 << 0)
+#define TIMER_PERIODIC    (1 << 1)
+#define TIMER_LOAD        (1 << 2)
+#define TIMER_INT_ENABLE  (1 << 3)
+#define TIMER_INT_PENDING (1 << 4)
+#define TIMER_CHAIN       (1 << 5)
+
+
+/* clang-format off */
+
+enum { timer1 = 0, timer2, timer3, timer4 };
+
+/* clang-format on */
+
+
+struct {
+	volatile u32 *timer0_base;
+	intr_handler_t handler;
+	volatile time_t jiffies;
+	spinlock_t sp;
+	u32 ticksPerFreq;
+} timer_common;
+
+
+static int _timer_irqHandler(unsigned int irq, cpu_context_t *ctx, void *data)
+{
+	volatile u32 st = *(timer_common.timer0_base + GPT_TCTRL1) & TIMER_INT_PENDING;
+
+	if (st != 0) {
+		++timer_common.jiffies;
+		/* Clear irq status */
+		*(timer_common.timer0_base + GPT_TCTRL1) |= TIMER_INT_PENDING;
+		*(timer_common.timer0_base + GPT_TCTRL1) &= ~TIMER_INT_PENDING;
```

Comment:
Comment why we set and clear this bit would be nice. Also, I wonder if we shouldn't place store barrier between this writes (and before leaving the handler too)?